chat.sympl

```diff
@@ -183,177 +228,260 @@ def _guard_input(input_description: str, input_: str) -> None:
     with PostTxArgs(room_channel):
         _invite_to_room_execute(room_channel, new_member)
-    cvm.job_start()
-
 
 @executable
-def _invite_to_room_execute(room_channel: ChannelName, new_member: KeyAlias) -> None:
+def _invite_to_room_execute(room_channel: ChannelName, new_member: KeyAlias) -> InviteToRoomEvent:
     room = _get_room(room_channel)
     members = room.members
 
-    if not std.contains_using(members, cvm.tx.key_alias, _str_eq):
-        not_a_member : str = cvm.tx.key_alias
-        cvm.error(f'Member {not_a_member} does not belong to the room. Operation denied.')
-
-    if not std.contains_using(members, cvm.tx.key_alias, _str_eq):
-            not_a_member : str = cvm.tx.key_alias
-            cvm.error(f'Member {not_a_member} does not belong to the room. Operation denied.')
-
+    #perform invite checks
+    invite_to_room_checks(room, new_member)
+    
+    #modify contract storage
     room.members = room.members + [new_member]
     cvm.storage.put(Identifier('room'), room)
+
+    #creates an invite to room event
     invite_to_room_event = InviteToRoomEvent(room=room, inviter=cvm.tx.key_alias, invitee=new_member)
     cvm.create_event('InviteToRoomEvent', std.json(invite_to_room_event))
-    cvm.job_complete(std.json(invite_to_room_event))
+    return invite_to_room_event
+
+#runs checks to see if the caller has the correct permissions to invite the 
+#a new member to a room
+@helper
+def invite_to_room_checks(room: Room, new_member: KeyAlias) -> None:
+    caller : KeyAlias = cvm.tx.key_alias
+
+    #check if the caller is an owner
+    if not std.contains_using(room.owners, caller, _str_eq):
+        cvm.error(f"{caller} is not an owner of the room {room.channel}.")
+
+    #check if the invitee is already a member
+    if std.contains_using(room.members, new_member, _str_eq):
+        cvm.error(f"Member {new_member} already in room {room.channel}.")
+
+    #check if the room is deleted
+    if room.is_deleted:
+        cvm.error(f"{room.channel} is deleted.")
 
-@public_clientside
+@clientside_helper
 def _remove_from_room(room_channel: ChannelName, member_to_remove: KeyAlias) -> None:
     room = _get_room(room_channel)
-    mrs : str = member_to_remove
-    if mrs == cvm.tx.key_alias:
-        cvm.error(f"Cannot remove self from room.")
-    members = room.members
-    if not std.contains_using(members, member_to_remove, _str_eq):
-        member_to_remove_str : str = member_to_remove
-        room_channel_str : str = room_channel
-        cvm.error(f'Member {member_to_remove_str} not in room {room_channel_str}.')
+    #run checks on removing
+    remove_from_room_checks(room, member_to_remove, cvm.tx.key_alias)
+
+    #if the member to remove is an owner, we want to remove ownership
+    #privileges, and then promote a new owner if there are no
+    #owners left
+    if std.contains_using(room.owners, member_to_remove, _str_eq):
+       _demote_owner(room_channel,member_to_remove)
```

Comment:
here and later on, there is a pattern that could potentially create issues. We are doing a secure channel membership operation in the clientside part and we are updating the record on storage in the executable. This means that a malicious actor could just run the executable and create a conflict between the record on storage and the actual channel membership. 
In general, all logical operation should be done inside the executable to avoid similar attacks.

Reply:
I'll add this check and action into the @ executable function as well.